core/trino-spi/src/main/java/io/trino/spi/connector/ConnectorMetadata.java

```diff
@@ -1029,9 +1030,13 @@ default Optional<LimitApplicationResult<ConnectorTableHandle>> applyLimit(Connec
      */
     default Optional<ConstraintApplicationResult<ConnectorTableHandle>> applyFilter(ConnectorSession session, ConnectorTableHandle handle, Constraint constraint)
     {
+        // applyFilter is expected not to be invoked with a "false" constraint
         if (constraint.getSummary().isNone()) {
             throw new IllegalArgumentException("constraint summary is NONE");
         }
+        if (FALSE.equals(constraint.getExpression())) {
+            throw new IllegalArgumentException("constraint expression is FALSE");
```

Comment:
I know who guarantees that `!constraint.getSummary().isNone()` (i can find it by checking the callers and arriving to PPITS)

Who guarantees this new condition? Document.

Reply:
Sorry I haven't documented yet because I waited for a response to https://github.com/trinodb/trino/pull/16206#discussion_r1150789956 
> There's no explicit check for that, but `DomainTranslator`'s implementation translates `FALSE` expressions into `TupleDomain.none()` (via `visitBooleanLiteral`) so the remaining expression shouldn't be `FALSE` and therefore the translated connector expression shouldn't be `FALSE` as well. Do you think I better add an explicit check within `PushPredicateIntoTableScan`?

For now, I've added a comment with this explanation.